src/information_theory/linear_code.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2022 Bolton Bailey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bolton Bailey
+-/
+import information_theory.hamming
+import linear_algebra.linear_independent
+import linear_algebra.affine_space.affine_subspace
+import linear_algebra.finite_dimensional
+import data.set.basic
+
+/-!
+# Linear Codes
+
+This file introduces Linear codes. A linear code is a type of error-correcting code defined as a
+linear subspace of a finite-dimensional vector space.
+
+## Main Definitions
+
+* `linear_code 𝓓 F`: The type of linear codes with domain `𝓓` over field `F`
+* `reed_solomon k D` : The code consisting of all polynomials of degree `≤ k` evaluated on a
+subset `D` of the field
+-/
+
+/--
+A linear error-correcting code, defined as a subspace of the vector space of functions from a
+domain into a field.
+-/
+def linear_code (𝓓 F : Type) [fintype 𝓓] [field F] := submodule F ( 𝓓 -> F )
+
+namespace linear_code
+
+variables {𝓓 F : Type} [fintype 𝓓] [field F]
+
+/-- The size of the domain of a code, i.e. the number of field elements transmitted -/
+def length (C : linear_code 𝓓 F) : ℕ := fintype.card 𝓓
+
+/-- The set of all valid codewords -/
+def codewords (C : linear_code 𝓓 F) := C.carrier
+
+/-- The dimension of the subspace of codewords -/
+noncomputable def dimension (C : linear_code 𝓓 F) : ℕ := set.finrank F C.codewords
+
+/--
+The minimum hamming distance between any two elements of the code. Equivalently, the minimum
+hamming distance of 0 from any nonzero element of the code.
+-/
+noncomputable def distance [decidable_eq F] (C : linear_code 𝓓 F) : ℕ :=
+Inf (set.image (λ w : hamming (λ i : 𝓓, F), hamming_dist w 0) (C.codewords \ {0}))
```

Comment:
Please use the `''` notation You can also use `nat.find` maybe.

Reply:
I've been working on this for a little while (actually longer than I probably should have been). Well done for submitting it. I'm not sure this is the right approach to the minimum distance, however - at the very least, there's a number of theorems you'll want to show are true about it. I have a branch called infsep where I've been trying to define this concept rigourously.